pkg/kafka/partition/reader.go

```diff
@@ -114,7 +154,251 @@ func (p *Reader) run(ctx context.Context) error {
+	}
+	if err := kerr.ErrorForCode(listRes.Topics[0].Partitions[0].ErrorCode); err != nil {
+		return 0, err
+	}
+
+	return listRes.Topics[0].Partitions[0].Offset, nil
+}
+
+// processNextFetchesUntilTargetOrMaxLagHonored process records from Kafka until at least the maxLag is honored.
+// This function does a best-effort to get lag below targetLag, but it's not guaranteed that it will be
+// reached once this function successfully returns (only maxLag is guaranteed).
+func (p *Reader) processNextFetchesUntilTargetOrMaxLagHonored(ctx context.Context, targetLag, maxLag time.Duration, recordsChan chan<- []Record) error {
+	logger := log.With(p.logger, "target_lag", targetLag, "max_lag", maxLag)
+	level.Info(logger).Log("msg", "partition reader is starting to consume partition until target and max consumer lag is honored")
+
+	attempts := []func() (currLag time.Duration, _ error){
+		// First process fetches until at least the max lag is honored.
+		func() (time.Duration, error) {
+			return p.processNextFetchesUntilLagHonored(ctx, maxLag, logger, recordsChan)
+		},
+
+		// If the target lag hasn't been reached with the first attempt (which stops once at least the max lag
+		// is honored) then we try to reach the (lower) target lag within a fixed time (best-effort).
+		// The timeout is equal to the max lag. This is done because we expect at least a 2x replay speed
+		// from Kafka (which means at most it takes 1s to ingest 2s of data): assuming new data is continuously
+		// written to the partition, we give the reader maxLag time to replay the backlog + ingest the new data
+		// written in the meanwhile.
+		func() (time.Duration, error) {
+			timedCtx, cancel := context.WithTimeoutCause(ctx, maxLag, errWaitTargetLagDeadlineExceeded)
+			defer cancel()
+
+			return p.processNextFetchesUntilLagHonored(timedCtx, targetLag, logger, recordsChan)
+		},
+
+		// If the target lag hasn't been reached with the previous attempt that we'll move on. However,
+		// we still need to guarantee that in the meanwhile the lag didn't increase and max lag is still honored.
+		func() (time.Duration, error) {
+			return p.processNextFetchesUntilLagHonored(ctx, maxLag, logger, recordsChan)
+		},
+	}
+
+	var currLag time.Duration
+	for _, attempt := range attempts {
+		var err error
+
+		currLag, err = attempt()
+		if errors.Is(err, errWaitTargetLagDeadlineExceeded) {
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if currLag <= targetLag {
+			level.Info(logger).Log(
+				"msg", "partition reader consumed partition and current lag is lower than configured target consumer lag",
+				"last_consumed_offset", p.committer.lastCommittedOffset,
+				"current_lag", currLag,
+			)
+			return nil
+		}
+	}
+
+	level.Warn(logger).Log(
+		"msg", "partition reader consumed partition and current lag is lower than configured max consumer lag but higher than target consumer lag",
+		"last_consumed_offset", p.committer.lastCommittedOffset,
+		"current_lag", currLag,
+	)
```

Comment:
Shouldn't this be inside an if `if currLag > targetLag`?

Reply:
No, it doesn't need to be as this line is only logged once the loop has exited. If currLag is <= targetLag, we return early on line 317.